opencv 4.2 dnn module build failed

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => 4.2.0
- Operating System / Platform => win 64
- Compiler => vs2015

##### Detailed description

<!-- your description -->
My desktop environment:  cuda 10.1, gtx 1080ti, cudnn 7.6.3
CMake entry `WITH_CUDNN`, `OPENCV_DNN_CUDA `were enabled.

Only the opencv_dnn project is build failed, and the other projects build success and libraries are generated.

Output error text :

> ...
4>  activations.cu
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(16): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getGridDim<0>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(17): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getGridDim<1>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(18): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getGridDim<2>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(21): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getBlockDim<0>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(22): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getBlockDim<1>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(23): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getBlockDim<2>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(26): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getBlockIdx<0>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(27): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getBlockIdx<1>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(28): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getBlockIdx<2>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(31): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getThreadIdx<0>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(32): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getThreadIdx<1>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda\grid_stride_range.hpp(33): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷unsigned int cv::dnn::cuda4dnn::csl::device::detail::getThreadIdx<2>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda4dnn\csl\cudnn/cudnn.hpp(41): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷cudnnDataType_t cv::dnn::cuda4dnn::csl::cudnn::detail::get_data_type<__half>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
4>d:\opencv\opencv4.2.0\sources\modules\dnn\src\cuda4dnn\csl\cudnn/cudnn.hpp(42): error C2912: 锟斤拷式专锟矫伙拷锟斤拷锟斤拷cudnnDataType_t cv::dnn::cuda4dnn::csl::cudnn::detail::get_data_type<float>(void)锟斤拷锟斤拷锟角猴拷锟斤拷模锟斤拷锟阶拷没锟?
...
...
4>  CMake Error at cuda_compile_1_generated_activations.cu.obj.Debug.cmake:279 (message):
4>    Error generating file
4>    D:/opencv/opencv4.2.0/build-contrib/modules/dnn/CMakeFiles/cuda_compile_1.dir/src/cuda/Debug/cuda_compile_1_generated_activations.cu.obj
4>
4>
========== 生成: 成功 3 个，失败 1 个，最新 14 个，跳过 0 个 ==========


How can I solve this problem?  Look forward to your reply.

--- a/modules/dnn/src/cuda/grid_stride_range.hpp
+++ b/modules/dnn/src/cuda/grid_stride_range.hpp
@@ -12,25 +12,29 @@
 namespace cv { namespace dnn { namespace cuda4dnn { namespace csl { namespace device {
 
     namespace detail {
-        template <int>  __device__ auto getGridDim()->decltype(dim3::x);
-        template <> inline __device__ auto getGridDim<0>()->decltype(dim3::x) { return gridDim.x; }
-        template <> inline __device__ auto getGridDim<1>()->decltype(dim3::x) { return gridDim.y; }
-        template <> inline __device__ auto getGridDim<2>()->decltype(dim3::x) { return gridDim.z; }
+        using dim3_member_type = decltype(dim3::x);
 
-        template <int> __device__ auto getBlockDim()->decltype(dim3::x);
-        template <> inline __device__ auto getBlockDim<0>()->decltype(dim3::x) { return blockDim.x; }
-        template <> inline __device__ auto getBlockDim<1>()->decltype(dim3::x) { return blockDim.y; }
-        template <> inline __device__ auto getBlockDim<2>()->decltype(dim3::x) { return blockDim.z; }
+        template <int>  __device__ dim3_member_type getGridDim();
+        template <> inline __device__ dim3_member_type getGridDim<0>() { return gridDim.x; }
+        template <> inline __device__ dim3_member_type getGridDim<1>() { return gridDim.y; }
+        template <> inline __device__ dim3_member_type getGridDim<2>() { return gridDim.z; }
 
-        template <int> __device__ auto getBlockIdx()->decltype(uint3::x);
-        template <> inline __device__ auto getBlockIdx<0>()->decltype(uint3::x) { return blockIdx.x; }
-        template <> inline __device__ auto getBlockIdx<1>()->decltype(uint3::x) { return blockIdx.y; }
-        template <> inline __device__ auto getBlockIdx<2>()->decltype(uint3::x) { return blockIdx.z; }
+        template <int> __device__ dim3_member_type getBlockDim();
+        template <> inline __device__ dim3_member_type getBlockDim<0>() { return blockDim.x; }
+        template <> inline __device__ dim3_member_type getBlockDim<1>() { return blockDim.y; }
+        template <> inline __device__ dim3_member_type getBlockDim<2>() { return blockDim.z; }
 
-        template <int> __device__ auto getThreadIdx()->decltype(uint3::x);
-        template <> inline __device__ auto getThreadIdx<0>()->decltype(uint3::x) { return threadIdx.x; }
-        template <> inline __device__ auto getThreadIdx<1>()->decltype(uint3::x) { return threadIdx.y; }
-        template <> inline __device__ auto getThreadIdx<2>()->decltype(uint3::x) { return threadIdx.z; }
+        using uint3_member_type = decltype(uint3::x);
+
+        template <int> __device__ uint3_member_type getBlockIdx();
+        template <> inline __device__ uint3_member_type getBlockIdx<0>() { return blockIdx.x; }
+        template <> inline __device__ uint3_member_type getBlockIdx<1>() { return blockIdx.y; }
+        template <> inline __device__ uint3_member_type getBlockIdx<2>() { return blockIdx.z; }
+
+        template <int> __device__ uint3_member_type getThreadIdx();
+        template <> inline __device__ uint3_member_type getThreadIdx<0>() { return threadIdx.x; }
+        template <> inline __device__ uint3_member_type getThreadIdx<1>() { return threadIdx.y; }
+        template <> inline __device__ uint3_member_type getThreadIdx<2>() { return threadIdx.z; }
     }
 
     template <int dim, class index_type = device::index_type, class size_type = device::size_type>

--- a/modules/dnn/src/cuda4dnn/csl/cudnn/cudnn.hpp
+++ b/modules/dnn/src/cuda4dnn/csl/cudnn/cudnn.hpp
@@ -37,9 +37,10 @@ namespace cv { namespace dnn { namespace cuda4dnn { namespace csl { namespace cu
         }
 
         /** get_data_type<T> returns the equivalent cudnn enumeration constant for type T */
-        template <class> auto get_data_type()->decltype(CUDNN_DATA_FLOAT);
-        template <> inline auto get_data_type<half>()->decltype(CUDNN_DATA_HALF) { return CUDNN_DATA_HALF; }
-        template <> inline auto get_data_type<float>()->decltype(CUDNN_DATA_FLOAT) { return CUDNN_DATA_FLOAT; }
+        using cudnn_data_enum_type = decltype(CUDNN_DATA_FLOAT);
+        template <class> cudnn_data_enum_type get_data_type();
+        template <> inline cudnn_data_enum_type get_data_type<half>() { return CUDNN_DATA_HALF; }
+        template <> inline cudnn_data_enum_type get_data_type<float>() { return CUDNN_DATA_FLOAT; }
     }
 
     /** @brief noncopyable cuDNN smart handle

